compiler/ir/ir.psi2ir/src/org/jetbrains/kotlin/psi2ir/generators/CallGenerator.kt

```diff
@@ -383,48 +384,41 @@ class CallGenerator(statementGenerator: StatementGenerator) : StatementGenerator
-
-        val valueArgumentsInEvaluationOrder = resolvedCall.valueArguments.values
-        val valueParameters = resolvedCall.resultingDescriptor.valueParameters
-
         val irBlock = IrBlockImpl(startOffset, endOffset, irResultType, IrStatementOrigin.ARGUMENTS_REORDERING_FOR_CALL)
 
-        val valueArgumentsToValueParameters = HashMap<ResolvedValueArgument, ValueParameterDescriptor>()
-        for ((index, valueArgument) in resolvedCall.valueArgumentsByIndex!!.withIndex()) {
-            val valueParameter = valueParameters[index]
-            valueArgumentsToValueParameters[valueArgument] = valueParameter
-        }
+        fun IrExpression.freeze(nameHint: String) =
+            if (isUnchanging())
+                this
+            else
+                scope.createTemporaryVariableInBlock(context, this, irBlock, nameHint).load()
 
-        val irArgumentValues = HashMap<ValueParameterDescriptor, IntermediateValue>()
+        irCall.dispatchReceiver = irCall.dispatchReceiver?.freeze("\$this")
+        irCall.extensionReceiver = irCall.extensionReceiver?.freeze("\$receiver")
 
-        for (valueArgument in valueArgumentsInEvaluationOrder) {
-            val valueParameter = valueArgumentsToValueParameters[valueArgument]!!
-            val irArgument = call.getValueArgument(valueParameter) ?: continue
-            val irArgumentValue = if (irArgument.hasNoSideEffects())
-                generateExpressionValue(irArgument.type) { irArgument }    // Computing a lambda has no side effects, can generate in place
-            else
-                scope.createTemporaryVariableInBlock(context, irArgument, irBlock, valueParameter.name.asString())
-            irArgumentValues[valueParameter] = irArgumentValue
+        val resolvedCall = call.original
+        val valueParameters = resolvedCall.resultingDescriptor.valueParameters
+        val valueArgumentsToIndex = HashMap<ResolvedValueArgument, Int>()
+        for ((index, valueArgument) in resolvedCall.valueArgumentsByIndex!!.withIndex()) {
+            valueArgumentsToIndex[valueArgument] = index
         }
-
-        resolvedCall.valueArgumentsByIndex!!.forEachIndexed { index, _ ->
-            val valueParameter = valueParameters[index]
-            irCall.putValueArgument(index, irArgumentValues[valueParameter]?.load())
+        for (valueArgument in resolvedCall.valueArguments.values) {
+            val index = valueArgumentsToIndex[valueArgument]!!
+            val irArgument = call.getValueArgument(valueParameters[index]) ?: continue
+            irCall.putValueArgument(index, irArgument.freeze(valueParameters[index].name.asString()))
         }
-
         irBlock.statements.add(irCall)
-
         return irBlock
     }
 }
 
-fun IrExpression.hasNoSideEffects() =
+fun IrExpression.isUnchanging() =
     this is IrFunctionExpression ||
             (this is IrCallableReference<*> && dispatchReceiver == null && extensionReceiver == null) ||
             this is IrClassReference ||
```

Comment:
note: technically those can have side effects due to class initialization, so maybe this is wrong